src/peft/utils/other.py

```diff
@@ -258,10 +258,50 @@ def _create_new_hook(self, old_hook):
         new_hook = old_hook_cls(**filtered_old_hook_attr)
         return new_hook
 
-    def forward(self, *args, **kwargs):
+    def _check_forward_args(self, x, *args, **kwargs):
+        """Check if the arguments are compatible with the configs and state of the model"""
+        adapter_names = kwargs.get("adapter_names", None)
+        if adapter_names is None:
+            return
+
+        if len(x) != len(adapter_names):
+            msg = (
+                "Length of `adapter_names` should be the same as the number of inputs, but got "
+                f"{len(adapter_names)} and {len(x)} respectively."
+            )
+            raise ValueError(msg)
+
+    def _mixed_batch_forward(
+        self, x: torch.Tensor, *args: Any, adapter_names: list[str], **kwargs: Any
```

Comment:
I think the issue of how to deal with the arguments is still unsolved. Right now, you're completely ignoring the `args`, which can lead to errors or even code that seems to pass but is actually incorrect. Say we have a module type with a forward function `forward(X, Y)`, where `X` and `Y` are tensors of the same size, as is, `Y` will just be dropped.

I think we have two options here:

1. We check all `args` and `kwargs` for tensors with the same length as `adapter_names` and those are being sliced. I.e. we make a best guess of what tensors are intended to be sliced and the rest is kept as is.
2. We restrict mixed batching to a handful of known modules whose signature we know and therefore can correctly handle the slicing. Most important would be `Linear`, then `Embedding`, `Conv2d` and transformers `Conv1D` would be nice to have.

Option 1 is a bit more flexible but option 2 more reliable and easier to implement. In fact, your current code is almost good for option 2, I'd just rename `x` to `input` and pass on `args`. Of course, we also need to add a type check to raise an error if the wrong module type is being used.

No matter what option is chosen, the restriction should be mentioned in the docs.